go/pserver/optimizer.go

```diff
@@ -1,42 +1,42 @@
 package pserver
 
 /*
-#include "optimizer.h"
+// TODO(zhihong): move compile flags to cmake go_library
+#cgo pkg-config: protobuf
+#cgo CFLAGS: -I ../../
+#cgo LDFLAGS: ../../build/paddle/optimizer/libpaddle_optimizer.a ../../build/proto/libpaddle_proto.a ../../third_party/install/glog/lib/libglog.a ../../third_party/install/gtest/lib/libgtest.a ../../third_party/install/gflags/lib/libgflags.a ../../third_party/install/openblas/lib/libopenblas.a -I/usr/local/lib/ -lprotobuf
+#include "paddle/optimizer/optimizer.h"
 */
 import "C"
 import (
 	"fmt"
 	"unsafe"
 )
 
-type optimizerType int
-
-const (
-	sgd optimizerType = iota
-)
-
 var nullPtr = unsafe.Pointer(uintptr(0))
 
 type optimizer struct {
 	opt *C.struct_paddle_optimizer
 }
 
-func newOptimizer(t optimizerType, learning_rate float64) *optimizer {
+func newOptimizer(paramWithConfigs ParameterWithConfig) *optimizer {
 	o := &optimizer{}
-	o.opt = C.paddle_create_SGD_optimizer(C.double(learning_rate))
+	p := paramWithConfigs.Param
+	c := paramWithConfigs.Config
+	o.opt = C.paddle_create_optimizer(C.uchar(c), C.int(len(c)), unsafe.Pointer(p.Content), c.int(p.Length), nullPtr, 0)
```

Comment:
`paddle_create_optimizer` takes the ownership of the `param_buffer` argument. So this does not work, because `p.Content` is a Go pointer, which is tracked by Go garbage collector, it will be garbage collected. We need to make a copy here.
For more pointer information please see: https://golang.org/cmd/cgo/#hdr-Passing_pointers

Another question is I don't get what is the benefit of changing the type of `p.Content` from `[]byte` to `*byte`:

- Before when `p.Content` have `[]byte` as type, in `paddle_send_grads` the `cArrayToSlice` function does not make a copy of the gradient buffer. It just use the buffer as underlying storage of the created slice. So changing the type to `*byte` does not improve performance.
- At this line of code, we have to make a copy anyway.

So I don't see any benefit of changing the type of `p.Content` from `[]byte` to `*byte`, it comes with added complexity: need a separate variable p.Length to track length, Go programmers are more familiar with slice (`[]byte`) than byte pointers (`*byte`).